In [1]:
%load_ext sql

In [2]:
con="postgresql://postgres:macbook@127.0.0.1/pagila"

In [3]:
%sql $con

'Connected: postgres@pagila'

<img src='files/facts.png'>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [4]:
%%sql select sum(sales_amount)
from factsales

 * postgresql://postgres:***@127.0.0.1/pagila
1 rows affected.


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [8]:
%%sql select d.country,sum(sales_amount)
from factsales f join dimcustomer d
on d.customer_key=f.customer_key
group by d.country

 * postgresql://postgres:***@127.0.0.1/pagila
108 rows affected.


[('Thailand', Decimal('419.04')),
 ('Virgin Islands, U.S.', Decimal('122.68')),
 ('Faroe Islands', Decimal('114.72')),
 ('Indonesia', Decimal('1510.33')),
 ('Bangladesh', Decimal('402.05')),
 ('Italy', Decimal('831.11')),
 ('Venezuela', Decimal('683.30')),
 ('Oman', Decimal('187.50')),
 ('Cameroon', Decimal('200.46')),
 ('Czech Republic', Decimal('133.71')),
 ('Sweden', Decimal('144.66')),
 ('Congo, The Democratic Republic of the', Decimal('212.50')),
 ('United Kingdom', Decimal('924.80')),
 ('Dominican Republic', Decimal('318.23')),
 ('Cambodia', Decimal('191.47')),
 ('Germany', Decimal('831.04')),
 ('Sri Lanka', Decimal('116.70')),
 ('American Samoa', Decimal('71.80')),
 ('Canada', Decimal('593.63')),
 ('Brunei', Decimal('113.65')),
 ('Finland', Decimal('101.74')),
 ('Anguilla', Decimal('106.65')),
 ('South Korea', Decimal('574.65')),
 ('Colombia', Decimal('709.41')),
 ('Saint Vincent and the Grenadines', Decimal('96.75')),
 ('Ukraine', Decimal('730.42')),
 ('Runion', Decimal('216.54')),
 ('Saudi Arabia', Decimal('523.79')),
 ('Argentina', Decimal('1434.48')),
 ('Liechtenstein', Decimal('114.72')),
 ('Latvia', Decimal('262.40')),
 ('North Korea', Decimal('113.69')),
 ('Holy See (Vatican City State)', Decimal('152.66')),
 ('Azerbaijan', Decimal('245.43')),
 ('Greece', Decimal('232.46')),
 ('Egypt', Decimal('694.39')),
 ('Afghanistan', Decimal('67.82')),
 ('Chad', Decimal('135.68')),
 ('Puerto Rico', Decimal('254.39')),
 ('India', Decimal('6630.27')),
 ('Iran', Decimal('950.75')),
 ('French Polynesia', Decimal('235.46')),
 ('Algeria', Decimal('383.10')),
 ('Chile', Decimal('328.29')),
 ('France', Decimal('374.04')),
 ('Gambia', Decimal('129.70')),
 ('Estonia', Decimal('115.70')),
 ('Slovakia', Decimal('89.74')),
 ('Vietnam', Decimal('746.28')),
 ('Israel', Decimal('422.01')),
 ('South Africa', Decimal('1204.15')),
 ('Nauru', Decimal('148.69')),
 ('Peru', Decimal('468.88')),
 ('Senegal', Decimal('100.75')),
 ('Kenya', Decimal('253.46')),
 ('Malaysia', Decimal('369.15')),
 ('Zambia', Decimal('134.67')),
 ('Kazakstan', Decimal('198.48')),
 ('Kuwait', Decimal('111.74')),
 ('Hong Kong', Decimal('142.70')),
 ('Madagascar', Decimal('93.78')),
 ('Yugoslavia', Decimal('259.43')),
 ('Japan', Decimal('3471.74')),
 ('Philippines', Decimal('2381.32')),
 ('Tuvalu', Decimal('120.74')),
 ('United States', Decimal('4110.32')),
 ('Iraq', Decimal('111.73')),
 ('Turkey', Decimal('1662.12')),
 ('Nigeria', Decimal('1511.48')),
 ('Mozambique', Decimal('344.20')),
 ('Switzerland', Decimal('252.39')),
 ('Ecuador', Decimal('390.13')),
 ('New Zealand', Decimal('92.76')),
 ('Hungary', Decimal('111.71')),
 ('China', Decimal('5802.73')),
 ('Belarus', Decimal('277.34')),
 ('Armenia', Decimal('118.75')),
 ('Pakistan', Decimal('525.72')),
 ('Russian Federation', Decimal('3045.87')),
 ('Netherlands', Decimal('586.66')),
 ('Romania', Decimal('235.38')),
 ('Paraguay', Decimal('275.38')),
 ('Brazil', Decimal('3200.52')),
 ('Austria', Decimal('307.22')),
 ('Bolivia', Decimal('183.53')),
 ('Ethiopia', Decimal('91.77')),
 ('Angola', Decimal('215.48')),
 ('Bahrain', Decimal('112.75')),
 ('Yemen', Decimal('510.83')),
 ('Lithuania', Decimal('73.76')),
 ('Bulgaria', Decimal('204.50')),
 ('Spain', Decimal('606.58')),
 ('Turkmenistan', Decimal('136.73')),
 ('Tunisia', Decimal('78.77')),
 ('Sudan', Decimal('227.46')),
 ('United Arab Emirates', Decimal('333.16')),
 ('Malawi', Decimal('123.72')),
 ('Morocco', Decimal('307.29')),
 ('Moldova', Decimal('127.66')),
 ('Myanmar', Decimal('194.48')),
 ('Mexico', Decimal('3307.04')),
 ('Tonga', Decimal('73.82')),
 ('Nepal', Decimal('116.78')),
 ('Tanzania', Decimal('341.17')),
 ('Taiwan', Decimal('1210.94')),
 ('Poland', Decimal('877.97')),
 ('French Guiana', Decimal('103.78')),
 ('Greenland', Decimal('137.66'))]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [9]:
%%sql select d.month,sum(sales_amount) as rev
from factsales f join dimdate d on d.date_key=f.date_key
group by d.month
order by d.month

 * postgresql://postgres:***@127.0.0.1/pagila
5 rows affected.


[(1, Decimal('4824.43')),
 (2, Decimal('9631.88')),
 (3, Decimal('23886.56')),
 (4, Decimal('28559.46')),
 (5, Decimal('514.18'))]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [10]:
%%sql
SELECT d.month, s.country, sum(sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY d.month, s.country
ORDER BY d.month, s.country, revenue desc;

 * postgresql://postgres:***@127.0.0.1/pagila
10 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10'))]

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

# All the above takes multiple scan of db which is not efficient so we go through all at once 

In [11]:
%%sql
SELECT d.month, s.country, sum(sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY grouping sets ((), d.month, s.country, (d.month, s.country))

 * postgresql://postgres:***@127.0.0.1/pagila
18 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Canada', Decimal('33689.74')),
 (None, 'Australia', Decimal('33726.77'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>

# CUBE 
- Group by CUBE (dim1, dim2, ..) , produces all combinations of different lenghts in one go.
- This view could be materialized in a view and queried which would save lots repetitive aggregations

## the cube is simillar to the groupingset where we specif the requirments in cube we dont do that 

In [12]:
%%time
%%sql
SELECT d.month, s.country, sum(sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY CUBE (d.month, s.country)

 * postgresql://postgres:***@127.0.0.1/pagila
18 rows affected.
CPU times: user 4.1 ms, sys: 3.1 ms, total: 7.21 ms
Wall time: 34.2 ms


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Canada', Decimal('33689.74')),
 (None, 'Australia', Decimal('33726.77'))]

## Revenue Total, by Month, by Country, by Month & Country All in one shot, NAIVE way
The naive way to create the same table as above is to write several queries and UNION them together. Grouping sets and cubes produce queries that are shorter to write, easier to read, and more performant. Run the naive query below and compare the time it takes to run to the time it takes the cube query to run.

In [13]:
%%time
%%sql
SELECT  NULL as month, NULL as country, sum(sales_amount) as revenue
FROM factSales
    UNION all 
SELECT NULL, dimStore.country,sum(sales_amount) as revenue
FROM factSales
JOIN dimStore on (dimStore.store_key = factSales.store_key)
GROUP by  dimStore.country
    UNION all 
SELECT cast(dimDate.month as text) , NULL, sum(sales_amount) as revenue
FROM factSales
JOIN dimDate on (dimDate.date_key = factSales.date_key)
GROUP by dimDate.month
    UNION all
SELECT cast(dimDate.month as text),dimStore.country,sum(sales_amount) as revenue
FROM factSales
JOIN dimDate     on (dimDate.date_key         = factSales.date_key)
JOIN dimStore on (dimStore.store_key = factSales.store_key)
GROUP by (dimDate.month, dimStore.country)

 * postgresql://postgres:***@127.0.0.1/pagila
18 rows affected.
CPU times: user 8.12 ms, sys: 3.03 ms, total: 11.2 ms
Wall time: 110 ms


[(None, None, Decimal('67416.51')),
 (None, 'Canada', Decimal('33689.74')),
 (None, 'Australia', Decimal('33726.77')),
 ('3', None, Decimal('23886.56')),
 ('5', None, Decimal('514.18')),
 ('4', None, Decimal('28559.46')),
 ('2', None, Decimal('9631.88')),
 ('1', None, Decimal('4824.43')),
 ('1', 'Australia', Decimal('2364.19')),
 ('1', 'Canada', Decimal('2460.24')),
 ('2', 'Australia', Decimal('4895.10')),
 ('2', 'Canada', Decimal('4736.78')),
 ('3', 'Australia', Decimal('12060.33')),
 ('3', 'Canada', Decimal('11826.23')),
 ('4', 'Australia', Decimal('14136.07')),
 ('4', 'Canada', Decimal('14423.39')),
 ('5', 'Australia', Decimal('271.08')),
 ('5', 'Canada', Decimal('243.10'))]